src/source/agc.rs

```diff
@@ -0,0 +1,175 @@
+use super::SeekError;
+use crate::{Sample, Source};
+use std::time::Duration;
+
+/// Constructs an `AutomaticGainControl` object with specified parameters.
+///
```

Comment:
for the novice like me, could you explain here the difference between gain, level and volume?

> Id suggest an attack_time of around 0.3 to 0.5 that seems to be a sweet spot for songs I tested.
> Also, target_level I adjusted to be 0.9 just to make sure it does not clip the audio; it was clipping for me at 1.0.

Its probably useful to suggest some values. Users can tweak them if they want they do not need to be perfect.

Reply:
### Gain
Gain is the amount of amplification applied to an audio signal. It is typically adjusted at the input stage of an audio device, such as a microphone preamp. Increasing the gain boosts the signal strength, which is particularly useful for capturing quieter sounds.

### Volume
Volume describes the perceived loudness of sound to the listener. It is typically controlled by the output of an audio device, such as speakers or amplifiers. While gain and level are more technical terms related to signal processing, volume is about how loud the sound is in the environment. Adjusting the volume affects how we perceive the sound without changing the underlying signal strength.

## The key difference
The key difference is that gain is about the strength of the audio signal, while volume is about the perceived loudness of the sound. Gain is adjusted at the input, while volume is controlled at the output.

So here i'm adjusting the audio strength in order to boost the entire signal in order to be able to hear those quiet sounds or to be able to hear the entire track if the entire track is very low on gain

### Regarding level:
Level refers to the strength of an audio signal at various points in the audio chain. It can be measured in decibels (dB) and indicates how strong the signal is, whether it's at the input of a mixer, the output of a track, or any other point in the processing chain. Levels can be adjusted to ensure that different audio elements work well together without distortion.